packages/teleterm/src/ui/services/quickInput/quickPickers.ts

```diff
@@ -130,14 +141,218 @@ export class QuickServerPicker extends ClusterPicker {
       serverUri: item.data.uri,
     });
   }
+
+  getAutocompleteResult() {
+    return null;
+  }
 }
 
 export class QuickCommandPicker implements QuickInputPicker {
+  launcher: CommandLauncher;
+  pickerRegistry: Map<string, QuickInputPicker>;
+
+  constructor(launcher: CommandLauncher) {
+    this.launcher = launcher;
+    this.pickerRegistry = new Map();
+  }
+
+  registerPickerForCommand(command: string, picker: QuickInputPicker) {
+    this.pickerRegistry.set(command, picker);
+  }
+
+  onPick(item: ItemCmd) {
+    this.launcher.executeCommand(item.data.name as any, null);
+  }
+
+  // TODO: Add tests.
+  // `tsh ssh foo` as input should return recognized-token.
+  // TODO(ravicious): Handle env vars.
+  getAutocompleteResult(input: string): AutocompleteResult {
+    const autocompleteCommands = this.launcher.getAutocompleteCommands();
+
+    // We can safely ignore any whitespace at the start.
+    input = input.trimStart();
+
+    // Return all commands if there's no input.
+    if (input === '') {
+      return {
+        kind: 'autocomplete.partial-match',
+        picker: this,
+        listItems: this.mapAutocompleteCommandsToItems(autocompleteCommands),
+      };
+    }
+
+    const matchingAutocompleteCommands = this.launcher
+      .getAutocompleteCommands()
+      .filter(cmd => {
+        const completeMatchRegex = new RegExp(`^${cmd.displayName}\\b`);
+        return (
+          cmd.displayName.startsWith(input) ||
+          // `completeMatchRegex` handles situations where the `input` akin to "tsh ssh foo".
+          // In that case, "tsh ssh" is the matching command, even though
+          // `cmd.displayName` ("tsh ssh") doesn't start with `input` ("tsh ssh foo").
+          completeMatchRegex.test(input)
+        );
+      });
```

Comment:
Technically the var definition above could be extracted to a separate function, but we're not repeating this job anywhere and I like being able to read code top-to-bottom. 😶